File: y2017/nim/day3.nim
import math
import strutils
import std/tables

const INPUT = 347991

const DIRS = [
  (0, 1), (-1, 0), (0, -1), (1, 0),
  (1, 1), (-1, 1), (-1, -1), (1, -1),
]

proc cellVal(n: int): (int, int) =
  var
    k = int(ceil((sqrt(float(n)) - 1) / 2))
    t = 2 * k + 1
    m = t * t
  
  dec t
  if n >= m - t:
    return (k - (m - n), -k)
  else:
    m -= t

  if n >= m - t:
    return (-k, -k + (m - n))
  else:
    m -= t

  return if n >= m - t: (-k + (m - n), k) else: (k, k - (m - n - t))

let cellPos = cellVal(INPUT)
let dist = cellPos[0].abs() + cellPos[1].abs()
echo "manhattan distance of cell #$1: $2" % [$INPUT, $dist]

var seen = {(0, 0): 1}.toTable
var at = 2
while true:
  let pos = cellVal(at)
  var posTotal = 0
  for d in DIRS:
    let n = (pos[0] + d[0], pos[1] + d[1])
    posTotal += seen.getOrDefault(n, 0)

  if posTotal > INPUT:
    echo "first value greater than $1: $2" % [$INPUT, $posTotal]
    break

  seen[pos] = posTotal
  inc at

File: y2017/rust/src/bin/day3.rs
use std::collections::HashMap;

const INPUT: u32 = 347991;
const DIRS: [(i32, i32); 8] = [
    (0, 1), (-1, 0), (0, -1), (1, 0),
    (1, 1), (-1, 1), (-1, -1), (1, -1),
];

fn cell_val(n: u32) -> (i32, i32) {
    // https://math.stackexchange.com/a/163101/713952
    let n = n as i32;
    let k = (((n as f64).sqrt() - 1.0) / 2.0).ceil() as i32;
    let mut t = 2 * k + 1;
    let mut m = t * t;
    t -= 1;

    if n >= m - t {
        return (k - (m - n), -k);
    } else {
        m -= t;
    }

    if n >= m - t {
        return (-k, -k + (m - n));
    } else {
        m -= t;
    }

    return if n >= m - t {
        (-k + (m - n), k)
    } else {
        (k, k - (m - n - t))
    };
}

fn main() {
    let cell_pos = cell_val(INPUT);
    let dist = cell_pos.0.abs() + cell_pos.1.abs();
    println!("manhattan distance of cell #{INPUT}: {dist}");

    let mut seen = HashMap::from([((0, 0), 1)]);
    let mut at = 2;
    loop {
        let pos = cell_val(at);
        let mut pos_total = 0;
        for d in DIRS {
            let n = (pos.0 + d.0, pos.1 + d.1);
            if seen.contains_key(&n) {
                pos_total += seen[&n];
            }
        }

        if pos_total > INPUT {
            println!("first value greater than {INPUT}: {pos_total}");
            break;
        }

        seen.insert(pos, pos_total);
        at += 1;
    }
}
